src/transport/raw/TCP.cpp

```diff
@@ -354,11 +354,40 @@ CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const Pe
             continue;
         }
 
-        // Buffer is incomplete if we reach this point
+        // If we reach this point, the head buffer does not contain a complete message.
+        // We may or may not have a complete message in a buffer chain.
         if (buffer->HasChainedBuffer())
         {
-            buffer->CompactHead();
-            continue;
+            if (messageSize > System::PacketBuffer::kMaxSizeWithoutReserve)
+            {
+                // This message will not fit in a maximum-size buffer.
+                if (messageSize <= buffer->TotalLength())
+                {
+                    // We have the oversize message in the current buffer chain.
+                    // Since it's oversize, it is definitely not a valid CHIP message, so we angrily toss it aside.
+                    buffer->ConsumeHead(messageSize);
+                    err = endPoint->AckReceive(messageSize);
+                    SuccessOrExit(err);
+                    err = CHIP_ERROR_MESSAGE_TOO_LONG;
+                    break;
+                }
+                // If we don't yet have the complete message, proceed along to open the receive window.
+            }
+            else
+            {
+                if (messageSize > buffer->MaxDataLength())
+                {
+                    // The current buffer head is too small for the message. We're gonna need a bigger buf.
+                    System::PacketBufferHandle newHead = System::PacketBufferHandle::NewWithData(
+                        buffer->Start(), buffer->DataLength(), static_cast<uint16_t>(messageSize - buffer->DataLength()));
+                    VerifyOrExit(!newHead.IsNull(), err = CHIP_ERROR_NO_MEMORY);
```

Comment:
I'm honestly a little torn about what we should do here.  If we never do the TCP-level ack for this data, then what?  Should we just go ahead and ack the receipt but drop the message?  Or will the caller handle the error here sanely somehow?

I'm really not sure what the semantics are of "the TCP-level send worked fine, but then we couldn't process the message at all" in terms of what reliable message delivery is supposed to guarantee.

Reply:
I would think that if you want the sender to ever send more bytes, you'll need to ack.  And now that I look at it, I'm seeing lots error paths where we don't do this.

Reply:
OK, I was having a little bit of a misunderstanding here.  I see now that `AckReceive` is the TCP-layer ack, and is a no-op unless the platform uses LWIP (everywhere else this happens automatically).  Now that that's cleared up, it seems very clear that the receiver must always ACK all bytes eventually.

> If we never do the TCP-level ack for this data, then what?

If we never do the TCP-level ack, the connection will be permanently stuck.  I think we need to verify that this code actually does ack in all cases, including error cases.  It does not appear to me that this is currently the case.

Reply:
@msandstedt Thank you for checking that!

In this PR we should at least not introduce new non-acking codepaths.  Fixing the existing cases can be a followup.